File: lib/dev_dock/container.rb
require 'docker'
require 'dev_dock/util'
require 'dev_dock/image'
require 'dev_dock/volumes'

module DevDock

	class DevContainer

		def initialize(image_name)
			@image = DevDock::DevImage.new(image_name)
			@volumes = DevDock::DevVolumes.new(@image)
			@name = DevDock::Util::snake_case("dev_dock_" + image_name)
		end

		def image
			@image
		end

		def volumes
			@volumes
		end

		def exist?
			Docker::Container.get(@name)
			true
		rescue Docker::Error::NotFoundError
			false
		end

		# kill container
		def kill
			Docker::Container.get(@name).kill		
		end

		def enable_x11(arguments)
			arguments.push '-v'
			arguments.push '/tmp/.X11-unix:/tmp/.X11-unix:ro'
			arguments.push '-e'
			arguments.push 'DISPLAY'
		end

		def run
			arguments = [
				'/usr/local/bin/docker',
				'run',
				'--privileged',
				'--name', @name,
				'--net=host',
				'--rm',
				'-ti',
				'--detach-keys',
				'ctrl-q,ctrl-q',
				'-u', `id -u`,
				'-e', 'GH_USER',
				'-e', 'GH_PASS',
				'-v', '/run/docker.sock:/var/run/docker.sock'
			]

			['workspaces', '.gitconfig', '.ssh'].each do |directory|
				arguments.push '-v', "#{ENV['HOME']}/#{directory}:/home/#{@image.user}/#{directory}"
			end

			if RUBY_PLATFORM == "x86_64-linux"
				enable_x11(arguments)
				arguments.push '-v', '/etc/localhost:/etc/localhost:ro'
			end

			@volumes.list.each do |volume|
				arguments.push '--mount', "source=#{volume.name},target=#{volume.path}"
			end

			arguments.push @image.name

			arguments.push 'tmux'
			arguments.push 'new'

			exec *arguments
		end
	end
end

File: slipway/container.py

import os
from .image import Image
from .volumes import Volumes
from .binds import Binds
from .util import snake_case


class Container(object):

    def __init__(self, client, args):
        self.client = client
        self.args = args
        self.image = Image(self.client, self.args)
        self.volumes = Volumes(self.client, self.args, self.image)
        self.binds = Binds(self.client, self.args, self.image)
        self.name = 'slipway_' + snake_case(self.args.image)

    def _run_arguments(self):
        arguments = [
            'docker',
            'run',
            '--net=host',
            '--rm',
            '-ti',
            '--detach-keys', 'ctrl-q,ctrl-q',
            '--name', self.name,
            '-e', 'GH_USER',
            '-e', 'GH_PASS',
            '-e', 'DISPLAY'
        ]

        for bind in self.binds.list():
            arguments.append('-v')
            argument = bind.host_path + ':' + bind.container_path
            if 'ro' in bind.type:
                argument += ':ro'
            arguments.append(argument)

        for volume in self.volumes.list():
            arguments.append('--mount')
            arguments.append(
                'source={},target={}'.format(volume.name, volume.path))
        arguments.append(self.image.name)
        arguments.append('tmux')
        arguments.append('new')
        return arguments

    def run(self):
        os.execvp('docker', self._run_arguments())
